pkg/promclient/multi_api.go

```diff
@@ -99,6 +106,33 @@ type MultiAPI struct {
 	antiAffinity    model.Time
 	metricFunc      MultiAPIMetricFunc
 	requiredCount   int // number "per key" that we require to respond
+	maxMem          int
+}
+
+// Starts a memory monitoring goroutine that oversees the memory usage of the process during a query and cancels memory intensive ones
```

Comment:
This doesn't seem to kill memory intensive queries, this seems to monitor memory usage and kill *all* queries associated with this context based on global process memory usage.


an example, assume we have 10 queries in-flight; they'll all be attached to this same multiAPI client; If 1 of those queries consumes tons of memory this will kill all 10 in-flight (which I'm not sure we can do better, but this is contrary to what this comment says it'll do). Similarly if we had multiple clients they'd all kill all of their inflight queries at the same time (more or less) since they are all measuring the global process usage.

Reply:
Looks like go profiling doesn't allow goroutine specific memory stats rather we're given process specific ones) which makes tracking memory usage for specific queries extremely hard. I left a comment there in case we find a better way to handle this.